### PR TITLE
fix cannot AST transform due to locals() not saving compile result #538 (2)

### PR DESCRIPTION
**Pull request by archibate**
Related issue id = #538

### CODE REVIEW

**Comment by archibate** [comment]
Reason given by `pydoc`:
```
locals()
    Return a dictionary containing the current scope's local variables.
    
    NOTE: Whether or not updates to this dictionary will affect name lookups in
    the local scope and vice-versa is *implementation dependent* and not
    covered by any backwards compatibility guarantees.
```

**Comment by archibate** [comment]
`node down: Not properly terminated` again.

**Comment by archibate** [comment]
Also note that this seems can make recursion functions failed to find themselves...
By the way, does taichi support recursion?

**Comment by archibate** [comment]
Don't merge, found bad case not covered by test:
```
@ti.func
def foo(x):
  return bar(x) * 2

@ti.func
def bar(x):
  return x + 100
```

**Comment by archibate** [comment]
Should compile func at runtime.

**Comment by yuanming-hu** [comment]
> By the way, does taichi support recursion?

Not yet. Actually we don't really have function calling support in Taichi IR: all functions are just inlined before generating the Taichi AST.

**Comment by archibate** [comment]
Sloved now. What's your time there?

**Comment by archibate** [comment]
> Not yet. Actually we don't really have function calling support in Taichi IR: all functions are just inlined before generating the Taichi AST.

We may support that in later version, any suggestions? GLSL for example, can have multiple functions other than `main()`, how about other platforms?

**Comment by yuanming-hu** [comment]
> Sloved now. What's your time there?

Awesome, thanks! It's 1pm here but I have a few meetings to participate. I'll take a look at your PR in a ~8 hours.

**Comment by archibate** [comment]
Fixed test, removd `@ti.func inc2`. Please use `@ti.classfunc` in class method.

**Comment by archibate** [comment]
Btw, I found GLSL doesn't support recursion for simplicity, see https://stackoverflow.com/questions/43601521/recursion-in-glsl-prohibited.

**Comment by yuanming-hu** [comment]
Interesting... To be honest I haven't found a use case in graphics that must use recursion :-)

**Comment by archibate** [comment]
eg. recursion can make ray tracing with reflection and refraction easier :)

File: tests/python/test_oop.py [hunk 1]
```diff
@@ -14,20 +14,11 @@ def __init__(self, n, m):
     @ti.classfunc
     def inc(self, i, j):
       self.val[i, j] += i * j
-      
-    @ti.func
```
Comment (yuanming-hu) [inline tests/python/test_oop.py]: Actually one thing that would make programmers' life easier is to merge `ti.func` and `ti.classfunc` together, and `ti.kernel` and `ti.classkernel` together. Then they don't have to use different decorators for kernels/functions inside/outside classes :-)

Comment (yuanming-hu) [reply tests/python/test_oop.py]: I managed to make `ti.func` work in classes, yet I failed to make `ti.kernel` work in classes. It'll be great if you can manage to do that :-) Then Taichi programmers will have one less concept to learn (`ti.classkernel`).

**Pull request merged**